kernel/src/engine/default/file_handler.rs

```diff
@@ -86,19 +90,64 @@ pub struct FileStream<F: FileOpener> {
     /// (before reaching the limit) and returns a batch iterator. If the file reader
     /// is not capable of limiting the number of records in the last batch, the file
     /// stream will take care of truncating it.
-    file_reader: F,
+    file_reader: Box<dyn FileOpener>,
     /// The stream state
     state: FileStreamState,
     /// Describes the behavior of the `FileStream` if file opening or scanning fails
     on_error: OnError,
 }
 
-impl<F: FileOpener> FileStream<F> {
+impl FileStream {
+    pub fn new_file_data_read_iterator<E: TaskExecutor>(
+        task_executor: Arc<E>,
+        schema: ArrowSchemaRef,
+        file_reader: Box<dyn FileOpener>,
+        files: &[FileMeta],
+        readahead: usize,
+    ) -> DeltaResult<FileDataReadResultIterator> {
+        let mut stream = FileStream::new(files.to_vec(), schema, file_reader)?;
+
+        // This channel will become the output iterator
+        // The stream will execute in the background, and we allow up to `readahead`
+        // batches to be buffered in the channel.
+        let (sender, receiver) = std::sync::mpsc::sync_channel(readahead);
+
+        let executor_for_block = task_executor.clone();
+        task_executor.spawn(async move {
+            while let Some(res) = stream.next().await {
+                let sender = sender.clone();
+                let join_res = executor_for_block
+                    .spawn_blocking(move || sender.send(res))
+                    .await;
+                match join_res {
+                    Ok(send_res) => match send_res {
+                        Ok(()) => continue,
+                        Err(_) => break,
+                    },
+                    Err(je) => {
+                        panic!("Couldn't join spawned task, runtime is likely in bad state: {je}")
+                    }
+                }
+            }
+        });
+
+        // Create a thread-safe iterator, because engine may consume it from multiple threads.
+        let it = receiver
+            .into_iter()
+            .map(|rbr| rbr.map(|rb| Box::new(ArrowEngineData::new(rb)) as _));
+        let it = std::sync::Mutex::new(it);
+        let it = std::iter::from_fn(move || match it.lock() {
```

Comment:
might be worth a comment that the closure is called _on each iteration step_, so we're not locked for the whole time.